Build GDNative client library

o01eg opened the pull request:
_Edit by @Vezzra: supersedes #3205 /edit_

Additional option to build GDNative library to use in Godot.

Adds option `BUILD_CLIENT_GODOT` (default **OFF**) and `GODOT_CPP_ROOT` to point at `godot-cpp` repository.

Exposes property `optionsDB` with methods to access FreeOrion's OptionsDB.

Exposes property `networking` with methods to connect to the server. Incoming messages are emits `ping` signal.

Exposes some incoming network messages as signals on main node include `game_start` message.

To test add 

```
p
GodotPlayer:p:1
otherPlayer2:p:1
```
into `auth.txt` config file and start the server with `./freeoriond --hostless -q --log-file /proc/self/fd/1 --setup.ai.player.count 0 --network.server.conn-human-empire-players.min 0 --network.server.unconn-human-empire-players.max 0`. This will start the game on the server and the only thing client need to do is to connect to the server.

To test single player game place godot binary into the same folder as freeoriond binary.

C++ sources of it got from https://github.com/Vezzra/fo-godot-prototype

Forum thread: https://freeorion.org/forum/viewtopic.php?f=9&t=11744

o01eg committed:
Initial GDNative support.

o01eg committed:
Initialize ClientApp from Godot

o01eg committed:
Parse resources

o01eg committed:
Expose optionsDB to access FreeOrion's OptionsDB.

o01eg committed:
Initial networking

o01eg committed:
Join the game and signal about incoming network messages

o01eg committed:
Send name of message to Godot

o01eg committed:
Initialize ClientApp and log system before check for default folder

o01eg committed:
Ignore .import files produced by Godot Editor

o01eg committed:
Add method to authorize on the server

o01eg committed:
Move auth_request to separate signal

o01eg committed:
Process important messages get started game

o01eg committed:
Start single player game

o01eg committed:
Get access to universe systems

Vezzra committed:
Add basic Godot prototype client app

Vezzra reviewed (approved):
So far so good. Only one comment, see below - I had to move the initialization of `GodotClientApp` back to where you originally placed it.

Otherwise I now can at least build the GDNative library and start the Godot client app. Still issues (see my post above), but it's a start. Now we need to start fixing and improving things... :wink:

Vezzra committed:
Set the build option for the experimental Godot client to ON by default

Vezzra committed:
Update .gitignore to exclude godot-cpp build dir

Vezzra committed:
Update .gitignore to exclude godot-cpp build artefacts

Vezzra committed:
Copy freeoriongodot library and dependencies directly to godot/client/bin

Vezzra committed:
GDNative library freeoriongodot now at res://bin

Vezzra committed:
Add GDNative freeoriongodot library and dependencies for macOS

Vezzra committed:
Move all Godot client .gitignore stuff into godot folder

Vezzra committed:
Make sure godot/client/bin folder exists before copying build artefacts to it

o01eg committed:
Place freeorion libraries depending on system

o01eg committed:
Fix relative resource path issues.

o01eg committed:
Initial Android support

Builds only libraries as Android supposed to used them from java
application.
Directories code isn't working but it's better than nothing.

o01eg committed:
Get JNIEnv to get access to Android directories

o01eg committed:
Cleanup unsupported single player game

o01eg committed:
Draw starlanes between stars.

o01eg committed:
Show fleets from the server

o01eg committed:
Use Godot threading for networking thread

o01eg committed:
Don't use removed global variables

o01eg committed:
Add GUT for unit tests

Run test with:
godot.x11.opt.tools.64 --verbose -d -s --path $PWD addons/gut/gut_cmdln.gd -gtest=res://test/integration/quickstart.gd

o01eg committed:
Update GUT to 7.1.0 to fix GDNative support

o01eg committed:
Test awaiting new single player game

o01eg committed:
Add support for headless server

o01eg committed:
Remove unneded libraries copy

o01eg committed:
Tweak godot client initialization

o01eg committed:
Add godot binaries for snap

o01eg committed:
Install freeoriongodot library to absolute constant path

o01eg committed:
Install FreeOrion Godot package

o01eg committed:
Get command-line arguments

o01eg committed:
Try to add FreeOrionGodot command support

o01eg committed:
Add methods for new turn and orderds

o01eg committed:
Export default content to Android apk

o01eg committed:
Try to fix snap packaging

o01eg committed:
Log missing resource directory

o01eg committed:
Try to change freeoriond detection

o01eg committed:
Fix update of background parsing

o01eg committed:
Attach dialogs to popup just before show

o01eg committed:
Rename networking thread variable

o01eg committed:
Add dialog window to connect to different servers

o01eg committed:
Add some error processing

o01eg committed:
Add password dialog if server requires authorisation

o01eg committed:
Show server address in log

o01eg committed:
Add Android rights to use Internet

o01eg committed:
Try to scale UI on mobile devices

o01eg committed:
Tune scale size

o01eg committed:
Revert "Remove copy to not yet decided path of Godot project"

This reverts commit d75fe5b303bd86d3a58c8e643e0afd23e7c4b2ef.
